sklearn/preprocessing/tests/test_encoders.py

```diff
@@ -690,9 +642,229 @@ def test_encoders_has_categorical_tags(Encoder):
+    'auto', [['a', 'b', None], [0, 2, np.nan]]
+])
+def test_ohe_missing_value_mixed_sanity(categories):
+    # sanity check for mixed data with missing values
+    X = np.array([['a', 'b', None], [0, np.nan, 2]], dtype=object).T
+    expected_X_trans = np.array([
+        [1, 0, 0, 1, 0, 0],
+        [0, 1, 0, 0, 0, 1],
+        [0, 0, 1, 0, 1, 0],
+    ])
+    ohe = OneHotEncoder(sparse=False, handle_unknown='ignore',
+                        categories=categories)
+    X_trans = ohe.fit_transform(X)
+    assert_allclose(expected_X_trans, X_trans)
+
+    X_test = np.array([
+        ['b', np.nan],
+        [None, 0],
+        ['c', 2],
+        ['a', 4]
+    ], dtype=object)
+    X_trans = ohe.transform(X_test)
+    expected_X_test = np.array([
+        [0, 1, 0, 0, 0, 1],
+        [0, 0, 1, 1, 0, 0],
+        [0, 0, 0, 0, 1, 0],
+        [1, 0, 0, 0, 0, 0],
+    ])
+
+    assert_allclose(expected_X_test, X_trans)
+
+
+@pytest.mark.parametrize("categories", [
+    'auto', [['cat', 'dog', None], [0, 3, 4, np.nan]]
+])
+def test_ohe_missing_value_support_pandas(categories):
+    # check support for pandas with mixed dtypes and missing values
+    pd = pytest.importorskip('pandas')
+    df = pd.DataFrame({
+        'col1': ['dog', 'cat', None, 'cat'],
+        'col2': np.array([3, 0, 4, np.nan], dtype=float)
+    }, columns=['col1', 'col2'])
+    expected_df_trans = np.array([
+        [0, 1, 0, 0, 1, 0, 0],
+        [1, 0, 0, 1, 0, 0, 0],
+        [0, 0, 1, 0, 0, 1, 0],
+        [1, 0, 0, 0, 0, 0, 1],
+    ])
+
+    ohe = OneHotEncoder(sparse=False, handle_unknown='ignore',
+                        categories=categories)
+    df_trans = ohe.fit_transform(df)
+    assert_allclose(expected_df_trans, df_trans)
+
+    expected_categories = [np.array(['cat', 'dog', None]),
+                           np.array([0, 3, 4, np.nan])]
+
+    assert_array_equal(ohe.categories_[0], expected_categories[0])
+    assert_array_equal(ohe.categories_[1][:-1], expected_categories[1][:-1])
+    assert np.isnan(ohe.categories_[1][-1])
+
+
+def test_ohe_missing_value_support_pandas_categorical():
+    # checks pandas dataframe with categorical features
+    pd = pytest.importorskip('pandas')
+    df = pd.DataFrame({
+        'col1': pd.Series(['c', 'a', np.nan, 'b', 'a'], dtype='category'),
+    })
+    expected_df_trans = np.array([
+        [0, 0, 1, 0],
+        [1, 0, 0, 0],
+        [0, 0, 0, 1],
+        [0, 1, 0, 0],
+        [1, 0, 0, 0],
+    ])
+
+    ohe = OneHotEncoder(sparse=False, handle_unknown='ignore')
+    df_trans = ohe.fit_transform(df)
+    assert_allclose(expected_df_trans, df_trans)
+
+    assert len(ohe.categories_) == 1
+    assert_array_equal(ohe.categories_[0][:-1], ['a', 'b', 'c'])
+    assert np.isnan(ohe.categories_[0][-1])
+
+
+def test_ohe_missing_values_both_missing_values():
+    # test both types of missing of missing values are treated as its own
+    # category
+    X = np.array([['a', 'b', None, 'a', np.nan]], dtype=object).T
```

Comment:
not sure if it's necessary but having ``np.nan`` twice would check the deduplication logic (again, I know it's covered in other tests as well).